plugin/src/automation/equivalence.ml

```diff
@@ -245,3 +247,60 @@ let equiv_proof env evd l =
  * Prove section and retraction
  *)
 let prove_equivalence env evd = twice_directional (equiv_proof env evd)
+
+
+let lookup_constant = Libnames.qualid_of_string %> Nametab.locate_constant
+
+let fresh_constant env evd const =
+  Evd.fresh_constant_instance env evd const |> Util.on_snd mkConstU
+
+type pre_adjoint = {
+  orn : lifting;
+  sect : Names.Constant.t;
+  retr0 : Names.Constant.t
+}
+
+(*
+ * Prepare quantification for an almost-adjoint equivalence, by pulling out
+ * all the type-quantified values (i.e., type parameters/indices) into a new
+ * local context before giving back an argument list for the adjunction lemmas.
+ *)
+let quantify_pre_adjunction env evd { orn; sect; retr0 } =
```

Comment:
I may tweak some of the code here when I refactor with the new library, but this looks good for now